tcp.c

```diff
@@ -425,38 +425,48 @@ tcp_connect(char *server)
 	int i;
 
 #ifdef IPv6
-
 	int n;
-	struct addrinfo hints, *res, *ressave;
+	struct addrinfo hints, *addr;
+	static struct addrinfo *res = NULL;
 	char tcp_port_rdp_s[10];
 
-	snprintf(tcp_port_rdp_s, 10, "%d", g_tcp_port_rdp);
+	/* Only look up the server address once. The official Windows
+	   2008 documentation states that the windows farm name should
+	   be a round-robin DNS entry containing all the terminal
+	   servers in the farm. When connected to the farm address, if
+	   we look up the address again when reconnecting (for any
+	   reason) we risk reconnecting to a different server in the
+	   farm. */
+	if (res == NULL)
+	{
+		snprintf(tcp_port_rdp_s, 10, "%d", g_tcp_port_rdp);
 
-	memset(&hints, 0, sizeof(struct addrinfo));
-	hints.ai_family = AF_UNSPEC;
-	hints.ai_socktype = SOCK_STREAM;
+		memset(&hints, 0, sizeof(struct addrinfo));
+		hints.ai_family = AF_UNSPEC;
+		hints.ai_socktype = SOCK_STREAM;
 
-	if ((n = getaddrinfo(server, tcp_port_rdp_s, &hints, &res)))
-	{
-		logger(Core, Error, "tcp_connect(), getaddrinfo() failed: %s", gai_strerror(n));
-		return False;
+		if ((n = getaddrinfo(server, tcp_port_rdp_s, &hints, &res)))
+		{
+			logger(Core, Error, "tcp_connect(), getaddrinfo() failed: %s", gai_strerror(n));
+			return False;
+		}
 	}
 
-	ressave = res;
+	addr = res;
```

Comment:
If this is a round robin setup then `res` will contain all the addresses, so this change will not improve things. What we need to save is the specific `struct addrinfo` record that we managed to `connect()` to.